### PR TITLE
Paper signatures

### PR DESCRIPTION
**Pull request by Mnemotechnician**
## About the PR
Adds signatures.

While holding a pen, you can alt-click a paper (or use the verb menu - right-click) to sign it. This adds a stamp with your name on it to the paper.

The name on the signature depends on your id if you have one (so agent id can be used to forge papers), or on your character (entity) name if you don't.

## Why / Balance
To sign boykisser faxes

## Technical details
Added a new system, tried to avoid touching upstream files (except the metadata one). During GetVerbsEvent<AlternativeVerb>, the system checks if the entity is holding a pen, and adds a verb to execute TrySignPaper. That method creates a stamp info with the entities' name on it and calls TryStamp from paper system. That's basically it.

## Media
https://cdn.discordapp.com/attachments/968983104662409244/1237455186239160330/weeee-2024-05-07_20.16.46.mp4?ex=663bb573&is=663a63f3&hm=082593f2f9b7ac04a9cef50c55766e4dd635b0bcbc9a92b2a919ca5cda00732a&

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
None

**Changelog**
:cl:
- add: You can now sign paper by alt-clicking it while holding a pen.

### CODE REVIEW

**Comment by Bonktrauma** [comment]
This is amazing

**Comment by rosieposieeee** [comment]
nice, i recall frontier having this and it was great there

File: Content.Server/DeltaV/Paper/SignatureSystem.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+public sealed class SignatureSystem : EntitySystem
+{
+    [Dependency] private readonly TagSystem _tagSystem = default!;
+    [Dependency] private readonly PaperSystem _paper = default!;
+    [Dependency] private readonly IdCardSystem _idCard = default!;
+    [Dependency] private readonly PopupSystem _popup = default!;
+    [Dependency] private readonly AudioSystem _audio = default!;
+
+    // The sprite used to visualize "signatures" on paper entities.
+    private static readonly string SignatureStampState = "paper_stamp-signature";
```
Comment (deltanedas) [inline Content.Server/DeltaV/Paper/SignatureSystem.cs]: should probably be const

File: Content.Server/DeltaV/Paper/SignatureSystem.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+namespace Content.Server.DeltaV.Paper;
+
+public sealed class SignatureSystem : EntitySystem
+{
+    [Dependency] private readonly TagSystem _tagSystem = default!;
+    [Dependency] private readonly PaperSystem _paper = default!;
+    [Dependency] private readonly IdCardSystem _idCard = default!;
+    [Dependency] private readonly PopupSystem _popup = default!;
+    [Dependency] private readonly AudioSystem _audio = default!;
```
Comment (deltanedas) [inline Content.Server/DeltaV/Paper/SignatureSystem.cs]: sort them alphabetically

File: Content.Server/DeltaV/Paper/SignatureSystem.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            Act = () =>
+            {
+                TrySignPaper(uid, args.User, component);
+            },
+            Text = Loc.GetString("paper-sign-verb"),
+            DoContactInteraction = true,
+            Priority = 10
+        };
+        args.Verbs.Add(verb);
+    }
```
Comment (deltanedas) [inline Content.Server/DeltaV/Paper/SignatureSystem.cs]: the verb could easily be predicted by moving this to shared

File: Content.Server/DeltaV/Paper/SignatureSystem.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    /// <summary>
+    ///     Tries add add a signature to the paper with signer's name.
+    /// </summary>
+    /// <param name="silent">Whether to avoid showing a popup if the paper cannot be signed.</param>
+    public bool TrySignPaper(EntityUid paper, EntityUid signer, PaperComponent? paperComp = null)
```
Comment (deltanedas) [inline Content.Server/DeltaV/Paper/SignatureSystem.cs]: use Entity<PaperComponent?> so its harder to pass wrong args

File: Content.Server/DeltaV/Paper/SignatureSystem.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        };
+
+        if (!paperComp.StampedBy.Contains(stampInfo) && _paper.TryStamp(paper, stampInfo, SignatureStampState, paperComp))
+        {
+            // Show popups and play a paper writing sound
+            var signedOtherMessage = Loc.GetString("paper-signed-other", ("user", signer), ("target", paper));
+            _popup.PopupEntity(signedOtherMessage, signer, Filter.PvsExcept(signer, entityManager: EntityManager), true);
+
+            var signedSelfMessage = Loc.GetString("paper-signed-self", ("target", paper));
+            _popup.PopupEntity(signedSelfMessage, signer, signer);
```
Comment (deltanedas) [inline Content.Server/DeltaV/Paper/SignatureSystem.cs:75]: may be wrong but im pretty sure theres a helper method for this exact case

File: Resources/Locale/en-US/deltav/paper/signature.ftl [hunk 6]
```diff
@@ -0,0 +1,5 @@
+paper-sign-verb = Sign
+
+paper-signed-other = {CAPITALIZE(THE($user))} places their signature on {THE($target)}.
```
Comment (deltanedas) [inline Resources/Locale/en-US/deltav/paper/signature.ftl]: wording it as `Urist McHands signs the paper` would sound better imo

**Review by deltanedas — CHANGES_REQUESTED** [review]

**Comment by ThataKat** [comment]
Maybe Cybersun pen should be able to forge and not base it off ID? If I stole someone's ID as a passenger and was suddenly able to write their signature, that would be odd. It would make sense if signature was innate to the person, but it would play into the Cybersun pen's whole gimmick if it could do forgeries. Maybe allow it to open a UI and type any name to be the signature?

File: Content.Server/DeltaV/Paper/SignatureSystem.cs (hunk 5 above)
Comment (Mnemotechnician) [reply Content.Server/DeltaV/Paper/SignatureSystem.cs:75]: There doesn't seem to be? Checked other systems (bible, paper, pointing), they do the same.

File: Content.Server/DeltaV/Paper/SignatureSystem.cs (hunk 3 above)
Comment (Mnemotechnician) [reply Content.Server/DeltaV/Paper/SignatureSystem.cs]: How do I do that? Paper components are different on the server and the client, and supposedly events do not support polymorphism, so I cannot subscribe on SharedPaperComponent.

Comment (deltanedas) [reply Content.Server/DeltaV/Paper/SignatureSystem.cs]: well if paper component isnt properly in shared upstream theres nothing that can be done realistically, dont worry about it

**Review by deltanedas — APPROVED** [review]
lgtm now

**Pull request merged**